Aboiding

McAwesome123 opened the pull request:
This PR adds a new "Boids" pattern.

The yokies will, for the most part, bunch together and move as a large group. Or occasionally, in a few smaller groups.

McAwesome123 committed:
Add Boids pattern as copy of Bubbles

McAwesome123 committed:
Add movement rules for boids

McAwesome123 committed:
Push away from the edges

(and tweak some parameters & stuff)

McAwesome123 commented:
This will probably need a fair bit of clean-up and/or optimization.
I'm also not 100% sure everything functions as intended.

pierce-smith1 commented on graphics.cpp:
```diff
@@ -72,3 +72,59 @@ GLubyte *Texture::data() const {
 
 	return texture_data;
 }
+
```
I try to avoid operator overloading as much as possible, since it obscures function calls and this makes telling what is _actually_ going on in some code much harder.

A vector type like this is probably one of the most sensible uses but I still dislike it.

McAwesome123 replied:
the main reason I added these is cause adding vectors together and multiplying them by a constant make sense to me and I just don't wanna write `get<X>(...) += ...; get<Y>(...) += ...;` every single time.

pierce-smith1 replied:
Oh yeah totally, you can still have convenience functions that hide all of that mess. I'm just saying, I would rather have:
```cpp
void add(Point &l, const Point &r);
Point sum(const Point &l, const Point &r);
```
instead of...
```cpp
Point &operator +=(Point &l, const Point &r);
Point operator +(Point &l, const Point &r);
```
I will admit that with the non-operator approach, the actual operation is a little less obvious because of the naming (it's not immediately clear if `add` or `sum` is the mutating operation, but it IS immediately clear which one of `+=` and `+` is the mutating operation. Still, that is why we have types.) But I still prefer it.

pierce-smith1 replied:
Hm, one thing I didn't immediately think about is because `Point` is not a real class, any functions we write for it have to be free, like the operators you made here. A free function that's just called `add` is really dumb. A free overload for the + operator is significantly less dumb.

Given that, here's what I'll say. I want to make some changes to the point class, which might come through in another PR, so for now, let's just leave things as you've made them here.

pierce-smith1 commented on graphics.cpp:
```diff
@@ -72,3 +72,59 @@ GLubyte *Texture::data() const {
+	return Point(x, y);
+}
+
+Point operator+=(Point &l, const Point &r) {
```
An actual nitpick, if these operators stick around: assignment operators should return a reference. It eliminates the potential for a useless copy and is just generally the idiomatic way to do it.

pierce-smith1 commented on spritecontrol.cpp:
```diff
@@ -391,6 +392,247 @@ std::map<PatternName, GlobalPlayer::MoveFunction> GlobalPlayer::move_functions {
+			}
+		}
+
+		std::map<Sprite *, Point> separation_velocity_changes;
```
These maps are probably hurting the performance the most. I don't have any super specific optimization plan but I think these eventually turn into static arrays, OR maybe disappear altogether (can we just apply the transformations now instead of storing them and applying them later?)

McAwesome123 replied:
I imagine these could probably be static arrays that's the same size as the sprites vector.
The main reason I'm not applying the transformations here is cause that'd mean later sprites read different values than earlier sprites. but whether that's actually an issue I'm not sure.

pierce-smith1 replied:
> static arrays that's the same size as the sprites vector.

That's tough because the sprites vector is not static, we don't know how big it is until runtime. So we'd have to overcommit to a static array that is big enough to hold what we think the upper bound amount of sprites is, which is shitty from a lot of different perspectives.

McAwesome123 replied:
ah, right, array size needs to be known at compile time.
perhaps just a vector then, but I don't know.

pierce-smith1 commented on spritecontrol.cpp:
```diff
@@ -391,6 +392,247 @@ std::map<PatternName, GlobalPlayer::MoveFunction> GlobalPlayer::move_functions {
+			double average_angle = get_angle(velocity[current_sprite->id()]);
+			Point average_pos = Point(current_final);
+
+			for (Sprite *other_sprite : *sprites) {		// dejil... i am sorry...
```
Im not sure how we can avoid a nested for loop, but we should think about it, since getting rid of it would be a massive performance win.

McAwesome123 replied:
It'd help a lot but I have no clue how to actually get rid of it.

pierce-smith1 commented on spritecontrol.cpp:
```diff
@@ -391,6 +392,247 @@ std::map<PatternName, GlobalPlayer::MoveFunction> GlobalPlayer::move_functions {
+		const static double STRETCH_RATIO = (double) (ctx->rect().bottom) / ctx->rect().right;
+		const static double SEPARATION_Y_RADIUS = (6.0 / (cfg[Cfg::SpriteCount] / 1.5 + 40.0)) * std::pow(SCREEN_SIZE / (1080LL * 1920LL) / 3.0 + 0.7, 1.1);
+		const static double SEPARATION_X_RADIUS = SEPARATION_Y_RADIUS * STRETCH_RATIO;
+		const static double VISION_Y_RADIUS = (10.0 / (cfg[Cfg::SpriteCount] / 15.0 + 20.0)) * std::pow(SCREEN_SIZE / (1080LL * 1920LL) / 3.0 + 0.7, 1.1);
```
Did I really hard code this shit for 1920x1080? what a disaster? I truly did not give a shit.

pierce-smith1 reviewed (commented):
Pretty much everything that really sucks about this is inherited from bubbles. I don't have much to say about the math right now since if math works then it works. But before this goes in let's try to clean up some of the mess bubbles made.

Also i read this code on my phone on a screen that can barely fit one word of text so there are probably nitpicks im missing.

pierce-smith1 commented on spritecontrol.cpp:
```diff
@@ -391,6 +392,247 @@ std::map<PatternName, GlobalPlayer::MoveFunction> GlobalPlayer::move_functions {
+
+		// Converts a vector to an angle in the range (-M_PI, M_PI].
+		auto get_angle = [](const Point &vector) {
+			double x = get<X>(vector);
```
Neat trick from C++17 I haven't used at all, but I use the corresponding feature in javascript all the time - you can do this:
```cpp
auto [x, y] = vector;
// or, better if you don't mutate the variables (as in this function)
const auto &[x, y] = vector;
```
This is called structured binding and it's a benefit we get from Point just being a std::pair. Cool!

McAwesome123 replied:
:lksix:

pierce-smith1 commented on spritecontrol.cpp:
```diff
@@ -391,6 +392,247 @@ std::map<PatternName, GlobalPlayer::MoveFunction> GlobalPlayer::move_functions {
+			double angle = atan2(y, x);
+
+			// i'm sure there will be absolutely nothing wrong with this
+			if (x >= 0) {
```
Why these if checks and -+pi adjustments?

McAwesome123 replied:
there is the possibility that I don't know how atan2 works

pierce-smith1 commented on spritecontrol.cpp:
```diff
@@ -391,6 +392,247 @@ std::map<PatternName, GlobalPlayer::MoveFunction> GlobalPlayer::move_functions {
+
+		// Wraps angles to be inside the range (-M_PI, M_PI]
+		auto wrap_angle = [](double angle) {
+			angle += M_PI;
```
This is some wack ass logic. Why not just
```c++
if (angle > pi) return angle - 2pi;
if (angle < -pi) return angle + 2pi;
return angle;
```
This is only half a rhetorical question because I'm not 100% sure the above works, but I feel like it should..?

McAwesome123 replied:
technically there'd be the issue of "if the angle is <-3pi or >3pi then it'd wouldn't be in the proper range" but I'm not sure if that'd occur with the current usage

pierce-smith1 replied:
Change the ifs into whiles then ;)

McAwesome123 replied:
I only read this now but that is what I've done

pierce-smith1 commented on spritecontrol.cpp:
```diff
@@ -391,6 +392,247 @@ std::map<PatternName, GlobalPlayer::MoveFunction> GlobalPlayer::move_functions {
+			return sqrt(x * x + y * y);
+		};
+
+		auto normalize_vector = [get_vector_magnitude](const Point &vector) {
```
Nitpick, when capturing from a lambda it pays to just use `[&]` instead of writing everything out explicitly. Plus, something like another lambda really should be captured by reference.

pierce-smith1 commented on spritecontrol.cpp:
```diff
@@ -391,6 +392,247 @@ std::map<PatternName, GlobalPlayer::MoveFunction> GlobalPlayer::move_functions {
+		const static double DESIRED_VELOCITY_RETURN_MULT = DEFAULT_FORCE_MULT * 0.3;	// How strongly to accelerate sprites towards their desired velocity
+
+		static std::map<Id, Point> velocity;
+		static std::map<Id, double> desired_velocity;
```
Nitpick, this should probably be called something like `desired_speed` to emphasize it is not a "velocity" in the sense that is used elsewhere in the code.

pierce-smith1 commented on spritecontrol.cpp:
```diff
@@ -391,6 +392,247 @@ std::map<PatternName, GlobalPlayer::MoveFunction> GlobalPlayer::move_functions {
+
+				Point other_final = Point(other_sprite->final<X>(), other_sprite->final<Y>());
+				Point diff = current_final - other_final;
+				get<Y>(diff) *= STRETCH_RATIO;
```
This STRETCH_RATIO shit is really dumb. What we should have is a coordinate system that actually makes sense rather than the stupid -1 to 1 but-its-actually-stretched-out-into-a-rectangle system we are using now. I guess we can just continue doing it for now but let's hope to remove it in the future.

McAwesome123 replied:
I absolutely agree.
And additionally, the fact that it goes from -1 to 1 instead of 0 to 1 massively fucked with me earlier cause I couldn't figure out why the fuck sprites on the bottom half and the left half of the screen were getting absolutely launched.

pierce-smith1 commented on spritecontrol.cpp:
```diff
@@ -391,6 +392,247 @@ std::map<PatternName, GlobalPlayer::MoveFunction> GlobalPlayer::move_functions {
+			Point diff = sprite.second - sprite_velocity;
+			diff *= ALIGNMENT_FORCE_MULT;
+
+			sprite_velocity = normalize_vector(sprite_velocity + diff) * magnitude;
```
This looks like an innocent variable assignment, but actually it modifies a collection! I would avoid doing stuff like this and just write out `velocity[sprite.first->id()] = whatever` explicitly. That is, if these collections even continue existing, which they probably won't.

McAwesome123 replied:
what exactly does that mean?

pierce-smith1 replied:
```cpp
std::vector numbers{1, 2, 3};
int &x = numbers[1];
x = 9; // Surprise! This looks like an innocuous assignment to a local variable, 
       // but actually it changes the value of numbers[1].
numbers[1] = 9; // Now this is much more obvious.
```
Generally surprises are a bad thing. It is surprising when an assignment changes something that is not even on the left hand side of the expression. This is where stupid bugs that eat an hour of your life come from.
As a general rule of thumb, you should avoid having a non-const reference as a local variable.

McAwesome123 replied:
I mean, to be fair, modifying the collection is _exactly_ what I'm trying to do here, but that's fair.

pierce-smith1 commented on spritecontrol.cpp:
```diff
@@ -391,6 +392,247 @@ std::map<PatternName, GlobalPlayer::MoveFunction> GlobalPlayer::move_functions {
+
+				if (dist < desired_separation[current_sprite->id()]) {
+					diff /= dist * dist / desired_separation[current_sprite->id()];
+					separation_velocity_changes[current_sprite] = diff;
```
Thoughts on optimization; what if instead of storing these changes in a map and applying them later, we just applied them right now? Same for the other adjustments being stored in maps. It would have a butterfly effect on the rest of the computations, but it probably won't be that bad because they are all very subtle changes.

pierce-smith1 commented on spritecontrol.cpp:
```diff
@@ -391,6 +392,247 @@ std::map<PatternName, GlobalPlayer::MoveFunction> GlobalPlayer::move_functions {
+		const static double EDGE_PUSH_FORCE_MULT = DEFAULT_FORCE_MULT * 1.5;			// How strongly to push sprites away from the edges
+		const static double DESIRED_VELOCITY_RETURN_MULT = DEFAULT_FORCE_MULT * 0.3;	// How strongly to accelerate sprites towards their desired velocity
+
+		static std::map<Id, Point> velocity;
```
Thoughts on optimization: we could avoid having these maps entirely if the velocity, desired separation, vision range, etc. were properties of the Sprites themselves. (Not literally on the `Sprite` type, since we don't want to pay the cost of this extra information everywhere when it's only used for this one function) That would require a decent architecture change, obviously, but I can envision some system that is able to transform regular Sprites into some custom type for each move function before it is called for the first time. That would actually go a long way for the other movement functions too! Probably out of scope for this PR.

pierce-smith1 commented on spritecontrol.cpp:
```diff
@@ -391,6 +392,247 @@ std::map<PatternName, GlobalPlayer::MoveFunction> GlobalPlayer::move_functions {
+			// there's probably a better way to do this but i can't be bothered figuring it out now
+		}
+
+		for (auto &sprite : cohesion_velocity_changes) {
```
More on structured binding: you can write the for loop like this:
```cpp
for (auto &[sprite, velocity_change] : alignment_velocity_changes) {
```
And then what used to be `sprite.first` becomes `sprite` and what used to be `sprite.second` becomes `velocity_change`. Basically, it gives you nicer variable names for free, which we should definitely use _if_ these for loops continue to exist.

pierce-smith1 commented on spritecontrol.cpp:
```diff
@@ -391,6 +392,247 @@ std::map<PatternName, GlobalPlayer::MoveFunction> GlobalPlayer::move_functions {
+			}
+
+			if (edge_push != Point(0.0, 0.0)) {
+				get<X>(edge_push) = pow(abs(get<X>(edge_push)), 1.0/2.5) * (get<X>(edge_push) < 0.0 ? -1.0 : 1.0);
```
Random thought: while having our Point type just be a std::pair is nice, I really have grown to hate the ergonomics of this `get<X>` crap, and really wish we could just do point.x and point.y instead. That is very possible but it's the scope of another PR.

pierce-smith1 reviewed (commented):
I read the code for real and I think I get what's going on. A bunch more scattered thoughts and nitpicks.

McAwesome123 committed:
Minor changes and optimizations (debug)

Reduced the amount of times maps are accessed by a lot. It's not perfect, but debug should now run a few times faster at higher sprite counts.

McAwesome123 committed:
debugging shit

McAwesome123 committed:
Fix velocity alignment

Fixed the velocity alignment causing yokies to preferentially go to the right side of the screen and bunch up there (i think).

McAwesome123 commented:
Might still need some slight changes like adjusting parameters, etc, but I'd say it's close enough to being ready now.

McAwesome123 committed:
Adjust behaviour

Slight behaviour adjustment to hopefully reduce how often yonkers perpetually cyclone.
Also, they will occasionally "scatter" for a short period of time (ignore other yokies, save for trying to not be inside them).

McAwesome123 committed:
Remove chance of ignoring pack

McAwesome123 committed:
More behaviour changes

McAwesome123 committed:
Remove some randomness and unnecessary stuff

Removed randomness for separation and vision
Removed scaling based on screen size

McAwesome123 committed:
Apply calculated forces immediately

Additionally,
- Made edge push radius constant
- Added a separate scale constant so that it's harder for you to accidentally divide by zero
- Fixed mistake where only one of the radii were being used?

McAwesome123 committed:
Trim more code and temp debug stuff

McAwesome123 committed:
throws you in the river

pierce-smith1 reviewed (approved):
This is in really good shape now and I think it's appropriate to ship. There may be some minor issues with it but we can pick those up later if they come up.
